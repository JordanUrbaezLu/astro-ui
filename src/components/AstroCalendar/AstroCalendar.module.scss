.astroCalendarContainer {
    display: flex;
    flex-direction: column;
    width: 500px;
}

.astroCalendarNavigationContainer {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
}

.astroCalendarNavigationMonth {
    font-family: Verdana, Geneva, Tahoma, sans-serif;
}

.astroCalendarHeaderContainer {
    display: flex;
    flex-direction: row;
}

.astroCalendarHeaderLabel {
    text-align: center;
    flex: 1 1 0px;
}

.astroCalendarBodyContainer {
    display: flex;
    flex-wrap: wrap;
}

.astroCalendarDay {
    text-align: center;
    flex: 1 0 13%;
    pointer-events: none;

    &:focus {
        background-color: red;
    }
}