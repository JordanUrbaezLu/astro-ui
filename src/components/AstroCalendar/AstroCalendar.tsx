import * as React from "react";
import styles from "./AstroCalendar.module.scss";
import { IoChevronBackSharp, IoChevronForwardSharp } from "react-icons/io5";
import {
  getCalendarDays,
  getNextDay,
  getNextWeek,
  getPreviousDay,
  getPreviousWeek,
} from "./astroCalendarUtilities";

export interface AstroCalendarProps {
  customDate?: Date;
}

const AstroCalendar: React.FC<AstroCalendarProps> = ({
  customDate = new Date(),
}) => {
  const [date, setDate] = React.useState<Date>(customDate);
  const [focusedDate, setFocusedDate] = React.useState<number>(0);
  const [days, setDays] = React.useState<Array<number | null>>([]);
  const daysRefs = React.useRef<Array<HTMLDivElement | null>>([]);

  const headerLabels = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];

  const increaseMonth = () => {
    setDate(new Date(date.getFullYear(), date.getMonth() + 1, date.getDate()));
  };

  const decreaseMonth = () => {
    setDate(new Date(date.getFullYear(), date.getMonth() - 1, date.getDate()));
  };

  React.useEffect(() => {
    setDays(getCalendarDays(date));
    setFocusedDate(date.getDate());
  }, [date]);

  React.useEffect(() => {
    daysRefs.current[focusedDate]?.focus();
  }, [focusedDate]);

  return (
    <div className={styles.astroCalendarContainer}>
      <div className={styles.astroCalendarNavigationContainer}>
        <IoChevronBackSharp
          tabIndex={0}
          onKeyDown={(event) => {
            if (event.key === "Enter") {
              decreaseMonth();
            }
          }}
          onClick={decreaseMonth}
        />
        <div className={styles.astroCalendarNavigationMonth}>
          {`${date.toLocaleString("en-US", {
            month: "long",
          })} ${date.getFullYear()}`}
        </div>
        <IoChevronForwardSharp
          tabIndex={0}
          onKeyDown={(event) => {
            if (event.key === "Enter") {
              increaseMonth();
            }
          }}
          onClick={increaseMonth}
        />
      </div>
      <div className={styles.astroCalendarHeaderContainer}>
        {headerLabels.map((label, index) => {
          return (
            <div className={styles.astroCalendarHeaderLabel} key={index}>
              {label}
            </div>
          );
        })}
      </div>
      <div className={styles.astroCalendarBodyContainer}>
        {days.map((day, index) => {
          if (day !== null) {
            return (
              <div
                className={styles.astroCalendarDay}
                tabIndex={day === date.getDate() ? 0 : -1}
                onKeyDown={(event) => {
                  const { code } = event;

                  const keyCodesToGet = {
                    ArrowUp: getPreviousWeek,
                    ArrowDown: getNextWeek,
                    ArrowLeft: getPreviousDay,
                    ArrowRight: getNextDay,
                  };

                  if (code in keyCodesToGet) {
                    setDate(keyCodesToGet[code]);
                  }
                }}
                key={index}
                ref={(el) => (daysRefs.current[day] = el)}
              >
                {day}
              </div>
            );
          } else {
            return <div className={styles.astroCalendarDay} key={index} />;
          }
        })}
      </div>
    </div>
  );
};

export default AstroCalendar;
