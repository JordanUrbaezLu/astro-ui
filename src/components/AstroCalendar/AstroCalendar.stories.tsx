import { ComponentStory, ComponentMeta } from "@storybook/react";
import AstroCalendar from "./AstroCalendar";

export default {
  title: "Components/AstroCalendar",
  component: AstroCalendar,
} as ComponentMeta<typeof AstroCalendar>;

const Template: ComponentStory<typeof AstroCalendar> = (args) => {
  return <AstroCalendar {...args} />;
};

export const Default = Template.bind({});
Default.args = {
  isOpen: true,
};

export const SelectedDate = Template.bind({});
SelectedDate.args = {
  isOpen: true,
  selectedDate: new Date(2019, 5, 9),
};
