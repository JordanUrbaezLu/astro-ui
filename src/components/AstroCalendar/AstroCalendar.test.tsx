import { render, screen } from "@testing-library/react";
import AstroCalendar from "./AstroCalendar";
import { axe } from "jest-axe";

describe("AstroCalendar", () => {
  const { container } = render(
    <AstroCalendar customDate={new Date(2019, 5, 9)} />
  );
  expect(container.firstChild).toMatchSnapshot();
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <AstroCalendar customDate={new Date(2019, 5, 9)} />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Props", () => {
  test("Should render custom date correctly.", () => {
    render(<AstroCalendar customDate={new Date(2019, 5, 9)} />);
    expect(screen.getByText("June 2019")).toBeInTheDocument();
  });
});
