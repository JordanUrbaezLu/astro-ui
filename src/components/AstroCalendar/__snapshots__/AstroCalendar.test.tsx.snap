// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  class="astroCalendarContainer"
>
  <div
    class="astroCalendarNavigationContainer"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      tabindex="0"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M328 112L184 256l144 144"
        fill="none"
        stroke-linecap="square"
        stroke-miterlimit="10"
        stroke-width="48"
      />
    </svg>
    <div
      class="astroCalendarNavigationMonth"
    >
      June 2019
    </div>
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      tabindex="0"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M184 112l144 144-144 144"
        fill="none"
        stroke-linecap="square"
        stroke-miterlimit="10"
        stroke-width="48"
      />
    </svg>
  </div>
  <div
    class="astroCalendarHeaderContainer"
  >
    <div
      class="astroCalendarHeaderLabel"
    >
      Sun
    </div>
    <div
      class="astroCalendarHeaderLabel"
    >
      Mon
    </div>
    <div
      class="astroCalendarHeaderLabel"
    >
      Tue
    </div>
    <div
      class="astroCalendarHeaderLabel"
    >
      Wed
    </div>
    <div
      class="astroCalendarHeaderLabel"
    >
      Thu
    </div>
    <div
      class="astroCalendarHeaderLabel"
    >
      Fri
    </div>
    <div
      class="astroCalendarHeaderLabel"
    >
      Sat
    </div>
  </div>
  <div
    class="astroCalendarBodyContainer"
  >
    <div
      class="astroCalendarDay"
    />
    <div
      class="astroCalendarDay"
    />
    <div
      class="astroCalendarDay"
    />
    <div
      class="astroCalendarDay"
    />
    <div
      class="astroCalendarDay"
    />
    <div
      class="astroCalendarDay"
    />
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      1
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      2
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      3
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      4
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      5
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      6
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      7
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      8
    </div>
    <div
      class="astroCalendarDay"
      tabindex="0"
    >
      9
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      10
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      11
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      12
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      13
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      14
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      15
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      16
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      17
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      18
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      19
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      20
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      21
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      22
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      23
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      24
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      25
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      26
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      27
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      28
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      29
    </div>
    <div
      class="astroCalendarDay"
      tabindex="-1"
    >
      30
    </div>
    <div
      class="astroCalendarDay"
    />
    <div
      class="astroCalendarDay"
    />
    <div
      class="astroCalendarDay"
    />
    <div
      class="astroCalendarDay"
    />
    <div
      class="astroCalendarDay"
    />
    <div
      class="astroCalendarDay"
    />
  </div>
</div>
`;
